flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/metrics/utils/MetricsPredicateProvider.java

```diff
@@ -0,0 +1,147 @@
+                // If the job type is streaming:
+                // Deployment is measures from start of first deployment
+                resultPredicate = deploymentPair -> true;
+            } else {
+                throw new RuntimeException("Illegal State Passed");
+            }
+        }
+
+        return resultPredicate;
+    }
+
+    /**
+     * Get the predicate which marks end of the metrics collection.
+     *
+     * @param semantic job type
+     * @param executionState current state of the job
+     * @param expectedDeployments Expected deployments count
+     */
+    public static Predicate<JobExecutionStatsHolder> getEndPredicate(
+            JobType semantic, ExecutionState executionState, Set expectedDeployments) {
+        Predicate<JobExecutionStatsHolder> resultPredicate;
+        if (semantic == JobType.BATCH) {
+            if (executionState == ExecutionState.INITIALIZING) {
+                resultPredicate =
+                        jobExecutionStats -> jobExecutionStats.getCompletedDeployments() > 0;
+            } else if (executionState == ExecutionState.DEPLOYING) {
+                resultPredicate =
+                        jobExecutionStatsHolder ->
+                                jobExecutionStatsHolder.getInitializingDeployments() > 0
+                                        || jobExecutionStatsHolder.getCompletedDeployments() > 0;
+            } else {
+                throw new RuntimeException("Illegal State Passed");
+            }
+        } else {
+            if (executionState == ExecutionState.INITIALIZING) {
+                resultPredicate =
+                        jobExecutionStatsHolder ->
+                                jobExecutionStatsHolder.getCompletedDeployments()
+                                        == expectedDeployments.size();
+            } else if (executionState == ExecutionState.DEPLOYING) {
+                resultPredicate =
+                        jobExecutionStatsHolder ->
+                                (jobExecutionStatsHolder.getInitializingDeployments()
+                                                + jobExecutionStatsHolder.getCompletedDeployments())
+                                        == expectedDeployments.size();
+            } else {
+                throw new RuntimeException("Illegal State Passed");
+            }
+        }
+
+        return resultPredicate;
+    }
+
+    /**
+     * Return the auxillary predicate involved in the metric start NOTE: The responsibility of
+     * passing in the correct input parameter lies with the caller.
+     *
+     * @param executionState Current state of the job
+     */
+    // NOTE: The responsibility of passing in the correct input parameter lies with the caller
+    public static Predicate<Pair<Integer, Integer>> getAuxillaryStartPredicate(
+            ExecutionState executionState) {
+        Predicate<Pair<Integer, Integer>> resultPredicate;
+        if (executionState == ExecutionState.DEPLOYING) {
```

Comment:
I remember that I suggested a generic implementation that bases everything on a triplet (preceding_states, target_states, following_states); why did you opt for hard-coding the predicates for each state? AFAICT this will make it more difficult to extend the set of metrics in the future (e.g., for runningTime or cancellingTime)